### PR TITLE
Enable OCSP stapling if certificate trust chain is provided

### PR DESCRIPTION
**Pull request by teohhanhui**
Previously disabled in 080a5157e6b1de3637e339a7cf54105f0316cfa7

### CODE REVIEW

**Comment by teohhanhui** [comment]
Possible issue:

> For a resolution of the OCSP responder hostname, the resolver directive should also be specified.

http://nginx.org/en/docs/http/ngx_http_ssl_module.html#ssl_stapling

But it's quite dangerous for us to set `resolver` because it affects all domain name resolving.

**Comment by kamermans** [comment]
This logic should be safe since it uses the system's resolvers if they are specified in `/etc/resolv.conf`.  If there are no `nameserver` lines, there is no DNS at all on the system, so Google's DNS servers are used.  In practice this can't normally happen in a Docker container since Docker itself falls back to Google's DNS if the host's `resolv.conf` doesn't contain accessible nameservers.

```
#!/bin/sh -e

RESOLVERS=$(awk '$1 == "nameserver" {print $2}' /etc/resolv.conf)

if [ "x$RESOLVERS" = "x" ]; then
    echo "Warning: unable to determine private DNS resolvers, using public DNS instead" >&2

    # Fallback to Google DNS servers
    RESOLVERS="8.8.4.4 8.8.8.8"
fi

echo resolve $RESOLVERS
```

To integrate this, we would need to export the `$RESOLVERS` variable to the system's environment -before- so it can be used for replacement in the `nginx.tmpl`.

**Comment by teohhanhui** [comment]
@kamermans In practice name resolution works just fine, but I'm not sure if it's static - that would be quite problematic.

**Comment by kamermans** [comment]
I'm not sure what you mean by static, but the resolver caches DNS lookups for the TTL of the record unless a different TTL is specified in the resolver line.

Perhaps you're concerned that without the resolver directive nginx will use the system resolver, so changes to the system DNS would be immediately active, but with resolver, it will use the specified DNS until a reload.   I can check the nginx source to see how it does name resolution by default.

Also, I see some documentation saying resolver is optional, can you confirm that it is required?

**Comment by kamermans** [comment]
The current Nginx-proxy config does not have the resolver directive, and according to the 2016 documentation, without resolver, Nginx uses the name servers from resolv.conf and permenantly caches the DNS responses:
> NGINX caches the DNS records until the next restart or configuration reload, ignoring the records’ TTL values.

https://www.nginx.com/blog/dns-service-discovery-nginx-plus/

This means that everything is basically already static, so using a method like mine above to populate the resolver directive is better than the default since it only caches DNS for the TTL.

**Comment by teohhanhui** [comment]
ping @jwilder

This should be ready for review.

**Comment by teohhanhui** [comment]
@kamermans The build is failing :(

**Comment by teohhanhui** [comment]
ping @jwilder

Can you have a look at this? Build should pass when jwilder/docker-gen#228 is merged.

**Comment by teohhanhui** [comment]
Build fixed. :smile:

File: docker-entrypoint.sh [hunk 1]
```diff
@@ -14,6 +14,13 @@ if [[ $DOCKER_HOST == unix://* ]]; then
 	fi
 fi
 
+# Compute the DNS resolvers for use in the templates
+export RESOLVERS=$(awk '$1 == "nameserver" {print $2}' ORS=' ' /etc/resolv.conf | sed 's/ *$//g')
+if [ "x$RESOLVERS" = "x" ]; then
+    echo "Warning: unable to determine DNS resolvers for nginx" >&2
+    unset RESOLVERS
```
Comment (teohhanhui) [inline docker-entrypoint.sh]: @kamermans This is okay, right?

Comment (kamermans) [inline docker-entrypoint.sh]: It is for now, yes, but at some point Docker said they are going to switch to a different resolver system, and that could cause issues, although I imagine they will maintain backwards compatibility.

**Comment by teohhanhui** [comment]
ping @jwilder

**Comment by teohhanhui** [comment]
@jwilder Is there anything blocking this? Any issues?

**Comment by teohhanhui** [comment]
Rebased.

Hope someone can help add some tests...

**Comment by jwilder** [comment]
Thanks @teohhanhui!

**Comment by buchdag** [comment]
Why the use of the `.crt` extension for the chain file ?

Isn't `.pem` supposed to be the standard extension for concatenated certificate container files ?

This is what jrcs/letsencrypt-nginx-proxy-companion [expect nginx-proxy to use](https://github.com/JrCs/docker-letsencrypt-nginx-proxy-companion/blob/master/app/letsencrypt_service#L37), meaning that currently OCSP stapling doesn't work out of the box with the LE companion.

This could also be fixed on `letsencrypt-nginx-proxy-companion` side but I'd prefer the template to expect a `.pem` for the chain file instead of a `.crt`.

Should I submit a PR ?

**Pull request merged**